lib/dataSource-recursiveFile.js

```diff
@@ -0,0 +1,166 @@
+(1) the top most level dir is read
+(2) the content of the dir is searched sequentially in the order returned by fs.readDir (generally alphabetically)
+(3) when a folder is encountered, it is searched in the order defined above (i.e. step 1-3 repeated with this new folder as root)
+(4) The first file that matches the given filename is returned
+
+Copyright 2017 Raising the Floor - International
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the License at
+https://github.com/fluid-project/kettle/blob/master/LICENSE.txt
+*/
+
+"use strict";
+
+var fluid = fluid || require("infusion"),
+    kettle = fluid.registerNamespace("kettle"),
+    fs = require("fs"),
+    path = require("path");
+
+fluid.require("querystring", require, "node.querystring");
+
+fluid.registerNamespace("kettle.dataSource");
+
+/**** RECURSIVE FILE DATASOURCE SUPPORT ****/
+
+fluid.defaults("kettle.dataSource.recursiveFile", {
+    gradeNames: ["kettle.dataSource"],
+    readOnlyGrade: "kettle.dataSource.recursiveFile",
+    invokers: {
+        getImpl: {
+            funcName: "kettle.dataSource.recursiveFile.handle",
+            args: ["{that}", "{arguments}.0", "{arguments}.1"] // options, directModel
+        }
+    }
+});
+
+kettle.dataSource.recursiveFile.rejectPromise = function (promise, errorMsg, statusCode) {
+    promise.reject({
+        isError: true,
+        message: errorMsg || "Unknown error occured attempting to handle recursiveFile dataSource",
+        statusCode: statusCode || 500
+    });
+};
+
+/**
+ * Asynchronously searches the filesystem recursively from the given `rootPath` for a file names `toFind`. Any file
+ * matching the filename is add to the `found` array. When the entire rootPath has been searched or if an
+ * error has occured, the callback function is called.
+ * @param rootPath {String} the root to search from. This should be  a valid folder.
+ * @param toFind {String} filename of the file to find
+ * @param found {Array} An array that will contain the results after the function is done
+ * @param callback {function} A function that has one parameter "error". The callback will be triggered when the
+ *    filesystem search has finished. If an error occured, it will be passed as a parameter to the function, else
+ *    nothing will be passed.
+ */
+kettle.dataSource.recursiveFile.findAll = function (rootPath, toFind, found, callback) {
+    fs.readdir(rootPath, function (err, files) {
+        if (err) {
+            callback(err.message, toFind);
+            return;
+        }
+
+        // set the number of pending async calls to the number of files, since we need to lstat each one
+        var pending = files.length;
+        if (pending === 0) { // empty folder
+            callback();
+            return;
+        }
+        var i = 0;
+
+        (function next() {
```

Comment:
Nuts and deeply unidiomatic!
This can all be vastly simpler than as implemented. We don't require for this to be backed by a filesystem which will change after startup - it is only ever applied for test data for unit or acceptance tests. We can simply do one grand scan at startup, build a table of which directory which file is in, and then for ever afterwards just look up that file in the table.

Reply:
This would mean that we would need to index the entire filesystem on startup though, wouldn't it? Since we dont know which folder the request is gonna be for, until the request is made (i.e. how would we know that it's the testData/* folders for example that are relevant, until we get a request for that... it might as well be the /tmp folder or the like)

Reply:
or am I missing something vital?

Reply:
Yes, that's right - we would index the filesystem tree under the mount point on startup